appinventor/components/src/com/google/appinventor/components/runtime/ChartDataModel.java

```diff
@@ -0,0 +1,114 @@
+import com.github.mikephil.charting.data.DataSet;
+import com.google.appinventor.components.runtime.util.YailList;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class ChartDataModel<T extends DataSet, D extends ChartData> {
+    protected D data;
+    protected T dataset;
+
+    /**
+     * Initializes a new ChartDataModel object instance.
+     *
+     * @param data  Chart data instance
+     */
+    protected ChartDataModel(D data) {
+        this.data = data;
+    }
+
+    /**
+     * Returns the size of the tuples that this Data Series
+     * accepts.
+     *
+     * @return  tuple size (integer)
+     */
+    protected abstract int getTupleSize();
+
+    public T getDataset() {
+        return dataset;
+    }
+
+    public ChartData getData() {
+        return data;
+    }
+
+    /**
+     * Changes the color of the data set.
+     *
+     * @param argb  new color
+     */
+    public void setColor(int argb) {
+        dataset.setColor(argb);
+    }
+
+    /**
+     * Changes the label of the data set.
+     *
+     * @param text  new label text
+     */
+    public void setLabel(String text) {
+        dataset.setLabel(text);
+    }
+
+    /**
+     * Sets the elements of the Data Series from a CSV-formatted String.
+     *
+     * @param elements String in CSV format
+     */
+    public void setElements(String elements) {
+        // Get the expected number of tuples
+        int tupleSize = getTupleSize();
+
+        // Split all the CSV entries by comma
+        String[] entries = elements.split(",");
+
+        // Iterate over every tuple (by grouping entries)
+        // We start from tupleSize - 1 since the (tupleSize - 1)-th
+        // entry will be the last entry of the tuple.
+        // The index is incremented by the tupleSize to move to the next
+        // group of entries for a tuple.
+        for (int i = tupleSize - 1; i < entries.length; i += tupleSize) {
+            List<String> tupleEntries = new ArrayList<String>();
+
+            // Iterate over all the tuple entries
+            // First entry is in (i - tupleSize + 1)
+            for (int j = tupleSize - 1; j >= 0; --j) {
+                int index = i - j;
+                tupleEntries.add(entries[index]);
+            }
+
+            // Add entry from the parsed tuple
+            addEntryFromTuple(YailList.makeList(tupleEntries));
+        }
+    }
+
+    /**
+     * Imports data from a YailList which contains nested tuples
+     *
+     * @param list  YailList containing tuples
+     */
+    public void importFromList(YailList list) {
+        // Iterate over all the tuples
+        for (int i = 0; i < list.size(); ++i) {
+            YailList tuple = (YailList)list.getObject(i);
+            addEntryFromTuple(tuple);
+        }
+    }
+
+    /**
+     * Adds an entry from a specified tuple.
+     * @param tuple  Tuple representing the entry to add
+     */
+    public abstract void addEntryFromTuple(YailList tuple);
```

Comment:
The core method to add entries is the addEntryFromTuple, which takes a YailList which is interpreted as a Tuple, and then parses it .

This allows for nice abstractions, because now, this method can be invoked no matter how many arguments one entry has.

So for LineCharts, for instance, a tuple is then (x, y)

If we had a BubbleChart (or a 3D Chart for that regard), the tuple would be (x, y, z)

Ideas on this approach? Is the overhead here significant, since we are storing values in a List and then parsing entries from it?